### PR TITLE
Dynamic board detection

### LINKED ISSUE #1091: Rewrite board detection and connection
At the moment, mobiflight scans for boards during startup once and then never again. All enumeration strategy is part of MobFlightCache, which holds all references to the connected MobiFlight Modules.

**Describe the solution you'd like**
I would like to change the way the board detection works in the following way:
* We have a mechanism that constantly scans for potential board connections
  * Serial Ports - run our existing routine to enumerate the serial ports
    * if a new serial port is found, raise a PortAvailable event
    * if a serial port was found before and is not avaialble anymore, raise a PortUnavailable event
  * On PortAvailable event, connect to the board and check whether it is compatible
    * if yes, then raise a new "MobiFlightBoardDetected" event
    * if no, then simply raise a "CompatibleBoardDetected" event
  * On "MobiFlightBoardDetected" event, check for the FirmwareVersion
    * if the FW version is outdated, raise a "FirmwareUpdatePossible" event
    * if the FW version is up-to-date, we don't have to raise anything special
  * On "MobiFlightBoardDetected" 
    * fetch the Board configuration
    * raise MobiFlightBoardConnected event

* On initial startup, we would still like to enumerate all present boards (i guess)
  * So once we receive the first "MobiFlightBoardConnected", we start a timer that will elapse after 1 sec
  * If we receiver another MobiFlightBoardConnected within a timeout, we wait for a potential next one.
  * If the timeout opccurs, we raise an new event "InitialModuleLookupFinished"

* we also change the code so that later after intial startup we can handle also the events of added or removed boards.
 
**Describe alternatives you've considered**
* Use USB HID Device connect events from Win... but I was not able to reliably process that... i had already tried. I think monitoring the Serial Ports like described above might also be more portable.

**Additional context**
https://github.com/MobiFlight/MobiFlight-Connector/issues/354

### PR DESCRIPTION
**Pull request by DocMoebiuz**
This PR rewrites the entire board detection and enumeration logic and makes it totally dynamic.

**Confirmations:**
- [x] All connected boards are detected on start up
- [x] Boards connected after MobiFlight start will be detected and are available for config thereafter
  - [x] board shows in the mobiflight boards settings dialog
  - [x] device config upload possible
  - [x] output / input configs show them in the list
- [x] Auto-firmware update and checks are working
  - [x] during start up
  - [x] for every connected device after start 
  - [x] auto-update still depends on the global setting
  - [x] when new boards are detected while the settings dialog is displayed, no auto-update is performed
- [x] Boards disconnected after Mobiflight will be removed from any list
  - [x] mobiflight boards settings dialog
  - [x] output / input configs 
- [x] Boards can be flashed
- [x] Board settings list updates dynamically (like device manager in windows)
- [x] i18n 
- [x] ~~documentation updated~~ there is no explicit documentation that needed to be updated.
- [x] Unit tests

Open issues:
- [ ] Ignored ports dissappear on re-connect.
- [ ] Ignored ports require reststart

**Notes:**

Improved board types & labels for boards without firmware installed
image

Renamed the labels for the options
image

fixes #1091 

related to #1149 - this fix improves the pro micro flashing because it can change ports after uploading the firmware or resetting. The reconnect logic automatically deals with it nicely. It still doesn't address the fact that pro micros should be flashed only in sequence. We have to monitor the ports and wait for the new port to show up.

### CODE REVIEW

File: MobiFlight/Monitors/UsbDeviceMonitor.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        /// as opposed to being connected as COM port.
+        /// </summary>
+        /// <returns>The list of connected USB drives supported by MobiFlight.</returns>
+        override protected void ScanPorts()
+        {
+            var result = new List<PortDetails>();
+            
+            foreach (var drive in DriveInfo.GetDrives())
+            {
+                // Issue 1074: Failing to check for IsReady caused an IOException on certain machines
```
Comment (neilenns) [inline MobiFlight/Monitors/UsbDeviceMonitor.cs:20]: The refactor of this into a new class is cool but it lost an important bug fix from the old code:

```C#
// Issue 1089: Network drives take *forever* to return the drive info slowing down startup. Only check removable drives.
                if (drive.DriveType != DriveType.Removable)
                {
                    Log.Instance.log($"Drive {drive.Name} ({drive.DriveType}) isn't a removable drive, skipping.", LogSeverity.Debug);
                    continue;
                }
```

**Review by neilenns — CHANGES_REQUESTED** [review]
Love the approach overall, super cool change!

**Comment by JaimeLeon2** [comment]
Sebastian,
Tested this build.  Installed in a new folder.  

It started looking very well.  I was able at first to connect and disconnect boards and I was amazed that MF Connector could correctly and automatically detect the boards going in and out of line.  Then I started MSFS, loaded the FBW A320 mcc and linked the config.  I started to test with the sim running.

When I tried to disconnect and reconnect a board after auto-loading the FBW A320 config.  Mobiflight crashed.  Something broke in my installation as the automatic detection of boards seem to disappear.  Something must have deactivate it.  I restarted Mobiflight several times and it detected all the boards connected at start, but failed to detect when I connected another board while running.  Also failed to detect when I disconnected a board while running.

I tried to reinstall the PR by copying the files over the existing files.  Same result.
I tried deleting the folder contents and recopying everything.  Same result.
I tried deleting the config file in Appdata\Local\Mobiflight.  After reinstalling the PR now I get the normal start after an update, but while loading the default mcc file, it crashes consistently.  No error message from Mobiflight.  It just quits.

I looked in Event Viewer and this is the message I get.  I tried deleting all the config files, but it keeps crashing.

Application: MFConnector.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.InvalidOperationException
   at System.ThrowHelper.ThrowInvalidOperationException(System.ExceptionResource)
   at System.Collections.Generic.Dictionary`2+ValueCollection+Enumerator[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].MoveNext()
   at MobiFlight.MobiFlightCache.Stop()
   at MobiFlight.ExecutionManager.Stop()
   at MobiFlight.ExecutionManager.ModuleCache_ModuleConnected(System.Object, System.EventArgs)
   at MobiFlight.MobiFlightCache+<SerialPortMonitor_PortAvailable>d__33.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__6_1(System.Object)
   at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()

**Comment by DocMoebiuz** [comment]
image

rename needed

**Pull request merged**